### DISCUSSION TITLE
Use of legal speed limits

### DISCUSSION THREAD

**Discussion by grutts** [valhalla/valhalla#4024] [Q&A]
Hello,

It appears that Valhalla can read truck speed limits by [parsing](https://github.com/valhalla/valhalla/blob/2f4d712e003c9d808a0cf370ff7789fa8814e2a6/src/mjolnir/pbfgraphparser.cc#L694) `maxspeed:hgv` from the map.

Can it do this for other vehicle types? For example, cars with trailers or buses (e.g. `maxspeed:bus`)

It appears this isn't currently possible from a search through the code and related issues, but I'd like to check to be sure.

If not, do you have another recommended method for loading in legal speed limits? We could run one instance of Valhalla for every vehicle type, which in turn uses a map that has been post-processed to replace the maxspeed with the corresponding vehicle's maxspeed on each road. However this is far from being an elegant solution.

Thanks in advance

**Comment by nilsnolde** [comment]
> Can it do this for other vehicle types? 

Currently we can't. Every separate speed will take up 7 bits and that's very expensive for us. So we only have normal `maxspeed` and `maxspeed:hgv` only.

Running one instance per profile seems a bit too much, but it's totally possible. With some code patches, you could turn the current `maxspeed:hgv` into any mode-specific `maxspeed`, but it requires a bit more diving into how it's currently used (e.g. in the `GraphTile::GetSpeed` function. Other than that, I also don't see any other way to support that currently.

**Reply by grutts** [reply]
OK thanks for confirming.

Another approach could be to parse `maxspeed:type` (e.g. `maxspeed:type=GB:nsl_single`) and apply logic to determine the appropriate speed limit on the road. An advantage being that it is more flexible to add new vehicle types, such as nuanced variations in the limits (e.g. based on vehicle length). I suppose the computational overhead could be acceptable? It would involve a single lookup on a mapping from `(type, profile)` to `maxspeed` per road.

**Comment by kevinkreiser** [comment]
is this a vanity metric you want to report back in the response or is it the actual limit needed to be used during the path calculation?

maybe im naive but I wonder what examples of this look like? differentiating between a truck speed and passenger vehicle speeds is the most salient example in real life that I know of, and even that is not so reliably tagged in osm.

reading the wiki https://wiki.openstreetmap.org/wiki/Key:maxspeed:type it also seemed that this tag doesn't tell you another speed for another mode it just tells you how the speed is governed on a particular road. to me this sounds like an enumeration not a speed with a mode. im not even sure what one would do with the enumeration value if one did store them in valhalla. the concrete example the wiki gives is pretty useless. it does seem though that in some cases when no regular maxspeed tag is given the enum could be used to look up the current legal speed for that country. but to me this should be done at parsing time and become the maxspeed value in the tiles rather than be stored separately.

anyway it's possible I didn't understand so apologies if that's the case.

**Reply by grutts** [reply]
The actual limit is needed during the path calculation. 

Here is an example using the speed limits in the UK (https://www.gov.uk/speed-limits) which I hope can clarify. A number of vehicles have a lower speed limit on single carriageways than cars - for example cars with trailers, motorhomes over 3 tonnes, and buses are all limited to 50 mph, instead of the 60 mph limit for cars. On motorways these vehicles are limited to 60 mph instead of the 70mph limit for cars. 

Not taking into account the speed limits on various roads leads odd routings. For example, setting a topspeed for such vehicles to 60mph but not taking into account that these vehicles can only drive 50mph on single carriageways, leads to routings which take the vehicle off the motorway and along smaller roads (because it would incorrectly appear that the vehicle can drive on a non-motorway road at the same speed as on the motorway, and the non-motorway path is a shorter distance).

**Reply by kevinkreiser** [reply]
ok but an important distinction is that these roads dont have 5 different speed limits listed on them. they have a sign that says this is a road which is subject to a specific law/rule. in other words its not like each road can have its own individual limit for each individual situation (mode of transport). its more like the road gets an additional type (enum) and you can use this to look up a speed depending on your mode of transport.

the reason i say this is because we already store a ton of information about the road in the graph. we know whether its a motorway or not and we really should know (but currently dont, due to OSM structure) whether or not its a single carriageway. we also know the admin (what country and state its in).

if we can codify the rules for different countries into something we just keep around in memory (seems totally doable imho) then we can simply look up the car with trailor speed if the edge is marked as adhering to these rules. this would be much cheaper than keeping a billion different speeds per mode on millions of edges, instead we just say you are a special case vehicle type and this edge has a bit marked that says it needs you to go look up the speed from the current laws.

now i still dont doubt that this will be a pain in the neck to implement. you'd need to enumerate maxspeed:type and maxspeed:source, hopefully there is only a few of those values. take a couple bits from directed edge to mark it. then you'd need to do the tedious work of turning the govt data into lookup tables, then you'd need to modify the costing to know about things like cars with trailors, and finally you'd need to update getspeed to care about it.

sounds horrible :smile: 

its certainly simpler to do the brute force approach you mentioned about but also, as you noted, terribly inefficient.

**Reply by grutts** [reply]
Sorry, I think we misunderstood each other slightly. I agree that storing the speeds for every vehicle type on the network doesn't scale. I think _we just say you are a special case vehicle type and this edge has a bit marked that says it needs you to go look up the speed from the current laws._ makes much more sense.

"you'd need to modify the costing to know about things like cars with trailors, and finally you'd need to update getspeed to care about it."
Does this part seem like a pain in the neck to implement? I can see that creating the lookup tables for every profile and country at once would be a huge pain, but I wonder whether building in the functionality to use a lookup table if available is smaller. Then the lookup tables can be built up incrementally over time.

**Reply by kevinkreiser** [reply]
oh yes i didnt mean we would need to implement every country right away. the lookup table should always have the option to say, i dont have an entry for that, in which case we fall back to the normal speed computation for that mode of travel. still though, in my view, the changes required to get all this working are not trivial, but indeed are manageable.
